src/mlpack/methods/ann/gan/cyclegan_impl.hpp

```diff
@@ -0,0 +1,613 @@
+    generatorY.ResetGradients(gradientGeneratorY);
+    generatorX.ResetGradients(gradientGeneratorX);
+    generatorY.Gradient(generatorY.parameter, 0, gradientGeneratorY, batchSize);
+    generatorX.Gradient(generatorX.parameter, 0, gradientGeneratorX, batchSize);
+  }
+
+  // Regenerate the original data back from generated data.
+  generatorX.Forward(std::move(generatedX));
+  generatorY.Forward(std::move(generatedY));
+  discriminatorY.predictors.cols(numFunctionsY, numFunctionsY + batchSize - 1) =
+      boost::apply_visitor(outputParameterVisitor, generatorX.network.back());
+  discriminatorX.predictors.cols(numFunctionsX, numFunctionsX + batchSize - 1) =
+      boost::apply_visitor(outputParameterVisitor, generatorY.network.back());
+  // TODO: Initialize discriminator responses
+
+  // Get the gradients from the re-generated data.
+  discriminatorY.Gradient(discriminatorY.parameter,
+      numFunctionsY, generatedGradientDY, batchSize);
+  discriminatorX.Gradient(discriminatorX.parameter,
+      numFunctionsX, generatedGradientDX, batchSize);
+  gradientDiscriminatorY += generatedGradientDY;
+  gradientDiscriminatorX += generatedGradientDX;
+
+  // Add cyclic loss.
+  currentInput = arma::mat(predictorsY.memptr() + (i * predictorsY.n_rows),
+      predictorsY.n_rows, batchSize, false, false);
+  res += lambda * norm(discriminatorY.predictors.cols(numFunctionsY,
+      numFunctionsY + batchSize - 1) - currentInput, 1);
+  currentInput = arma::mat(predictorsX.memptr() + (i * predictorsX.n_rows),
+      predictorsX.n_rows, batchSize, false, false);
+  res += lambda * norm(discriminatorX.predictors.cols(numFunctionsX,
+      numFunctionsX + batchSize - 1) - currentInput, 1);
+
+  // Update the Generator gradients.
+  if (currentBatch % generatorUpdateStep == 0 && preTrainSize == 0)
+  {
+    // Minimize (D(G(X)) - 1)^2 and (D(G(Y)) - 1)^2.
+    // Pass the error from Discriminators to Generators.
+    // TODO: Initialize discriminator responses (else this can be optimized)
+    discriminatorX.Gradient(discriminatorX.parameter, numFunctionsX,
+        generatedGradientDX, batchSize);
+    discriminatorY.Gradient(discriminatorY.parameter, numFunctionsY,
+        generatedGradientDY, batchSize);
+    generatorY.error = boost::apply_visitor(deltaVisitor,
+        discriminatorX.network[1]);
+    generatorX.error = boost::apply_visitor(deltaVisitor,
+        discriminatorY.network[1]);
+
+    generatorY.Predictors() = generatedY;
+    generatorX.Predictors() = generatedX;
+    generatorY.ResetGradients(generatedGradientGY);
+    generatorX.ResetGradients(generatedGradientGX);
+    generatorY.Gradient(generatorY.parameter, 0, generatedGradientGY, batchSize);
+    generatorX.Gradient(generatorX.parameter, 0, generatedGradientGX, batchSize);
+
+    gradientGeneratorY += generatedGradientGY;
+    gradientGeneratorX += generatedGradientGX;
+    gradientGeneratorY *= multiplier;
+    gradientGeneratorX *= multiplier;
+  }  
+
+  counterX++;
```

Comment:
@zoq This is not required internally anywhere, but it helps in keeping track of the mini-batch number being processed at the current time. I don't think it would lead to a wrong result, because if the batch size changes, the `counterX * batchSize >= numFunctionsX` will regulate itself.